### ISSUE TITLE
Moving code to Python

### ISSUE THREAD

**Issue by wesbarnett** [wesbarnett/snap-pac#39]
This is an issue created to let the community know that I am in the process of moving the bash script to Python. This is because I have grown increasingly frustrated with updating the bash script to include new features. bash is not a programming language so it's difficult to easily do the things I want quickly. Additionally with Python I can add unit tests which are currently missing.

Overall, the same functionality will be present in the Python version of the script. Obviously there will now be a dependency on python. One notable change will be that the configuration files will be merged into a single configuration file with a slightly different format. I know that will be an annoyance to some, but I believe it is worth the power and flexibility I will gain in being able to modify the code base.

**Comment by NicoHood** [comment]
I have mixed feelings about that. I've done lots of coding in several programming languages, also in bash and python.

I can understand that bash is really, really aweful sometimes, as you never know when the script might break. You can use shellcheck to avoid common mistakes, but there are still cases where bash is just super magic. If it gets too complex, it will be more likely to break. The good thing is, that it works on every system without any dependency. You can call the desired programs directly and not via ugly system calls.

Now with python the question is, does it require lots of system calls or are there python modules available? If they are available, the question is how good they are. I've tried git and gnupg integration, and the overall experience was very bad, so I switched from bash to python and back to back with my GPGit script. That just as a note. Also if those modules are not available in the official repos, you will get into trouble installing them.

Now this should not be a problem, since I am maintaining snap-pac for Archlinux [community]. But if a python dependency is not in the repos yet, I tend to avoid packaging so many more new dependencies, in case they are not so often used. Also I have the struggle on convincing their maintainers to add gpg signatures, but thats another story. But as always: It depends.

Since we are on an archlinux focussed package, the whole discussion is a bit more simple, as we dont need to add support for other distributions. Archlinux also has a quite good python integration for pacman, at least from what I remember. But it wasnt documented so well at that time.

tl;dr;
Now the conclusion is, if snap-pac will only depend on python and no more other python package I'd say go for it. If there are other dependencies, I'd say: It depends.

Does that help?

**Comment by wesbarnett** [comment]
I've got a working version in the master branch. No outside dependencies besides python. Nothing changes as far as packaging except for adding `python` as a dependency.

For the user there will be a single configuration file in the Python `ini` format. There's an example checked in to the repository.

I'm still testing this version out. We now have unit tests to test most of the functionality. But, I was hitting some errors sometimes on the post hook. Going to keep testing until I know it's been resolved;

```
[2021-02-27T08:11:33-0500] [ALPM] running 'zy_snapper-post.hook'...
[2021-02-27T08:11:33-0500] [ALPM-SCRIPTLET] /bin/sh: -c: line 1: unexpected EOF while looking for matching `"'
[2021-02-27T08:11:33-0500] [ALPM-SCRIPTLET] /bin/sh: -c: line 2: syntax error: unexpected end of file
[2021-02-27T08:11:33-0500] [ALPM-SCRIPTLET] ==> root: 
```

**Comment by wesbarnett** [comment]
Pre-release created with this change. https://github.com/wesbarnett/snap-pac/releases/tag/3.0.beta0

**Comment by NicoHood** [comment]
I guess you have a PKGBUILD for 3.0, would you like to share it with me?

**Comment by wesbarnett** [comment]
Sure, here it is:

```
pkgname=snap-pac
pkgver=3.0
pkgrel=1
pkgdesc="Pacman hooks that use snapper to create pre/post btrfs snapshots like openSUSE's YaST"
arch=('any')
url="https://github.com/wesbarnett/snap-pac"
license=('GPL')
depends=('snapper' 'pacman' 'btrfs-progs' 'python')
makedepends=('python-sphinx')
checkdepends=('python-pytest')
conflicts=('snap-pac')
source=("${url}/releases/download/${pkgver}/${pkgname}-${pkgver}.tar.gz"
         "${url}/releases/download/${pkgver}/${pkgname}-${pkgver}.tar.gz.sig")
validpgpkeys=("F7B28C61944FE30DABEEB0B01070BCC98C18BD66")
sha512sums=('a25aad8d2d529c918d220908657183b7316a0e8a9439b0bcff4e5ced06e2e65632a39f0debce225369c88a0295e78794a74a19c9f711233f4380245569ed560b'
            'SKIP')

check() {
    cd "${pkgname}"
    make test
}

package() {
    cd "${pkgname}"
    make DESTDIR="${pkgdir}" install
}
```

**Comment by NicoHood** [comment]
I have a remark on the config. I'd like to install it to `/etc/snap-pac.ini` instead of `/etc/snap-pac.ini.example`.

The config is fully commented, so it is no problem to have it there. In order to avoid overwrites on an upgrade we can use:
```
backup=('etc/snap-pac.ini')
```
As we did with all previous versions.

I think there was a note about the configuration file change. I'd like to add some text to people upgrading to >3.0, so they know their config no longer will take effect. Do you have a dedicated text/website for that notice?

This is what I'd use:
```
_3.0_1_changes() {
    echo ":: Config file format changed with version 3.0."
    echo "   Please adapt your config from /etc/snap-pac.conf to /etc/snap-pac.ini"
    echo "   https://wesbarnett.github.io/snap-pac/configuration.html"
}

post_upgrade() {
    local v upgrades=(
      3.0-1
    )

    for v in "${upgrades[@]}"; do
      if [[ $(vercmp "$v" "$2") -eq 1 ]]; then
        "_${v//-/_}_changes"
      fi
    done
}
```

**Comment by wesbarnett** [comment]
3.0.1 has the config file at `/etc/snap-pac.ini` instead of `/etc/snap-pac.ini.example`. The release notes for 3.0+ address the config format changes.

**Comment by NicoHood** [comment]
Thanks!
Why was the install hook removed here?

https://github.com/wesbarnett/snap-pac/commit/5271109cc4066415f058c70e28c5cbe136a9d183#diff-a9fa9fda98d7712cf2c04ce6206e406c554572bedfeac6e441f13410980f7f82

It seems to be not related to the commit message itself, so was that removal correct?

**Comment by wesbarnett** [comment]
> Thanks!
> Why was the install hook removed here?
> 
> [5271109#diff-a9fa9fda98d7712cf2c04ce6206e406c554572bedfeac6e441f13410980f7f82](https://github.com/wesbarnett/snap-pac/commit/5271109cc4066415f058c70e28c5cbe136a9d183#diff-a9fa9fda98d7712cf2c04ce6206e406c554572bedfeac6e441f13410980f7f82)
> 
> It seems to be not related to the commit message itself, so was that removal correct?

Yes that removal is correct. All that hook did was print a message to the screen indicating that a post snapshot would not be taken upon installation of snap-pac. That behavior has now changed - a single snapshot is taken if no pre snapshot is available like on installation, and the script itself logs that message instead of having a hook do it. I should have been more explicit in the commit message.

**Comment by NicoHood** [comment]
Thanks! I've updated the arch package accordingly.
